routing/src/interface.rs

```diff
@@ -0,0 +1,363 @@
+            self.oper_state = state;
+            // Todo: log change
+        }
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Set the administrative state of an interface
+    //////////////////////////////////////////////////////////////////
+    pub fn set_admin_state(&mut self, state: IfState) {
+        if self.admin_state != state {
+            self.admin_state = state;
+            // Todo: log change
+        }
+    }
+    // Todo: decide if set_oper_state() and set_admin_state() will be
+    // the ones to trigger the enforcement of the new state
+    // (e.g. by requesting the interface manager to bring up/down
+    // interfaces in kernel), or, instead, they just update the
+    // interface object seen by the rest of the routing system.
+
+    //////////////////////////////////////////////////////////////////
+    /// Attach an interface to a VRF. It is assumed that both the
+    /// interface and the VRF exist.
+    //////////////////////////////////////////////////////////////////
+    pub fn attach(&mut self, vrf: &Arc<Vrf>) -> Result<(), RouterError> {
+        if let Some(exist_vrf) = &self.vrf {
+            if !Arc::ptr_eq(exist_vrf, vrf) {
+                Err(RouterError::AlreadyAttached)
+            } else {
+                Ok(())
+            }
+        } else {
+            self.vrf = Some(vrf.clone());
+            Ok(())
+        }
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Detach an interface from its VRF (whichever it is)
+    //////////////////////////////////////////////////////////////////
+    pub fn detach(&mut self) {
+        self.vrf.take();
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Get the VRF that an interface is attached to, or None otherwise
+    //////////////////////////////////////////////////////////////////
+    pub fn get_vrf(&self) -> Option<&Vrf> {
+        self.vrf.as_deref()
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Add (assign) an IP address to an interface
+    //////////////////////////////////////////////////////////////////
+    pub fn add_ifaddr(&mut self, ifaddr: &IfAddress) {
+        self.addresses.insert(*ifaddr);
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Del (unassign) an IP address from an interface
+    //////////////////////////////////////////////////////////////////
+    pub fn del_ifaddr(&mut self, ifaddr: &IfAddress) {
+        self.addresses.remove(ifaddr);
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Tell if an interface has a certain IP address assigned
+    /// (regardless of the mask)
+    //////////////////////////////////////////////////////////////////
+    pub fn has_address(&self, address: &IpAddr) -> bool {
+        for (addr, _) in &self.addresses {
+            if addr == address {
```

Comment:
Why not use [`contains`](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.contains)?

Reply:
I think that the method was meant to find if the interface had a certain address configured (no matter the mask). If we'd use contains(), we would need to add the masklen, since the key is (address, len), right?
Now, if you asked me the utility of that function, .... that's another story :sweat_smile:

Reply:
oh, quite right.  I clearly didn't read that method correctly